src/gridcoin/upgrade.cpp

```diff
@@ -472,66 +701,73 @@ bool Upgrade::ExtractSnapshot()
         // if the zip file has no entries.
         if (!totaluncompressedsize)
         {
-            ExtractStatus.SnapshotZipInvalid = true;
+            ExtractStatus.SetSnapshotZipInvalid(true);
 
             LogPrintf("Snapshot (ExtractSnapshot): Error - snapshot.zip has no entries");
 
-            return false;
+            return;
         }
 
         // Now extract
-        for (i = 0; i < (uint64_t)entries; i++)
+        for (u_int64_t i = 0; i < entries; ++i)
         {
             if (zip_stat_index(ZipArchive, i, 0, &ZipStat) == 0)
             {
                 // Does this require a directory
                 if (ZipStat.name[strlen(ZipStat.name) - 1] == '/')
+                {
                     fs::create_directory(ExtractPath / ZipStat.name);
-
+                }
                 else
                 {
+                    struct zip_file* ZipFile;
+
                     ZipFile = zip_fopen_index(ZipArchive, i, 0);
 
                     if (!ZipFile)
                     {
-                        ExtractStatus.SnapshotExtractFailed = true;
+                        ExtractStatus.SetSnapshotExtractFailed(true);
 
                         LogPrintf("Snapshot (ExtractSnapshot): Error opening file %s within snapshot.zip", ZipStat.name);
 
-                        return false;
+                        return;
                     }
 
+
                     fs::path ExtractFileString = ExtractPath / ZipStat.name;
 
                     FILE* ExtractFile = fsbridge::fopen(ExtractFileString, "wb");
 
                     if (!ExtractFile)
                     {
-                        ExtractStatus.SnapshotExtractFailed = true;
+                        ExtractStatus.SetSnapshotExtractFailed(true);
 
                         LogPrintf("Snapshot (ExtractSnapshot): Error opening file %s on filesystem", ZipStat.name);
 
-                        return false;
+                        return;
                     }
 
-                    sum = 0;
+                    int64_t sum = 0;
 
-                    while ((uint64_t)sum != ZipStat.size)
+                    while ((uint64_t) sum < ZipStat.size)
                     {
+                        int64_t len = 0;
+                        char Buf[256*1024];
+
                         boost::this_thread::interruption_point();
 
-                        len = zip_fread(ZipFile, Buf, 1024*1024);
+                        len = zip_fread(ZipFile, &Buf, 256*1024);
 
```

Comment:
Believe it or not, this is the actual bug on the mac. It blows up on a 1M char buffer size. You have to lower it to 256*1024. I can't believe how long it took me to find this...